functions/src/modules/gmail.functions.ts

```diff
@@ -0,0 +1,240 @@
+  async (req: functions.Request, res: functions.Response) => {
+    const functionsURL = `https://${req.headers.host}/Gmail-oauth2callback`;
+    const oauth2ClientWithRedirect = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      functionsURL
+    );
+
+    const { tokens } = await oauth2ClientWithRedirect.getToken(req.query.code);
+    oauth2ClientWithRedirect.setCredentials(tokens);
+    await saveTokens(tokens);
+    res.send('Ok');
+  }
+);
+
+const fetchToken = async () => {
+  const queryResults = await db.ref(`/gmail/coordinator`).once('value');
+  const values = queryResults.val();
+
+  // https://github.com/GoogleCloudPlatform/cloud-functions-gmail-nodejs/blob/master/lib/oauth.js#L62
+  // https://github.com/googleapis/google-auth-library-nodejs/releases/tag/v2.0.0
+  if (
+    !values.oauth.expiry_date ||
+    // moment().valueOf() = Unix timestamp in milliseconds + 1 minute
+    values.oauth.expiry_date < moment().valueOf() + 60000
+  ) {
+    oauth2Client.credentials.refresh_token =
+      oauth2Client.credentials.refresh_token || values.oauth.refresh_token;
+    const result = await oauth2Client.getAccessToken();
+    await saveTokens(result.res.data);
+    return {
+      ...values,
+      oauth: result.res.data,
+    };
+  }
+  return values;
+};
+
+const storeHistoryIdInDatabase = async (historyId: any) => {
+  return await db.ref(`/gmail/coordinator`).update({
+    lastSyncHistoryId: historyId,
+  });
+};
+
+const saveLabelId = (labelId: string) => {
+  const newGmailTokensRef = db.ref(`/gmail/coordinator`);
+  newGmailTokensRef.update({
+    labelId,
+  });
+};
+
+// Fetch labelId from database if exists if not fetch from API and save in database
+const fetchLabelId = async (gmailClient: any): Promise<string> => {
+  const result = await db
+    .ref(`/gmail/coordinator/labelId`)
+    .once('value');
+  let labelId = result.val();
+
+  if (!labelId) {
+    const labelResults = await gmailClient.users.labels.list({ userId: 'me' });
+    labelId = labelResults.data.labels.filter(label => {
+      return label.name === functions.config().gmail.label_name;
+    })[0].id;
+    saveLabelId(labelId);
+  }
+
+  return labelId;
+};
+
+export const initWatch = functions.https.onRequest(
+  async (req: functions.Request, res: functions.Response) => {
+    // Initiate gmail client
+    const { oauth } = await fetchToken();
+    oauth2Client.setCredentials({
+      access_token: oauth.access_token,
+      refresh_token: oauth.refresh_token,
+    });
+    const gmail = await Google.google.gmail({
+      version: 'v1',
+      auth: oauth2Client,
+    });
+
+    try {
+      const labelId: string = await fetchLabelId(gmail);
+      const watchResults = await gmail.users.watch({
+        userId: 'me',
+        requestBody: {
+          labelIds: [labelId],
+          topicName: `projects/${
+            process.env.GCLOUD_PROJECT
+          }/topics/gmail-labeled-done`,
```

Comment:
Topic name should be taken from a constant.

Reply:
`const DONE_LABEL = 'gmail-labeled-done';`